### ISSUE TITLE
Channel.isWritable always returns false

### ISSUE THREAD

**Issue by yogeshvedpathak** [netty/netty#5920]
Hi, I upgraded netty from netty-4.0.30 to netty-4.0.41 because I was interested in getting fix for issue https://github.com/netty/netty/pull/5592. However after upgrading i ran into a serious issue. The entire IO request hangs because channel.isWritable returns false after sending some data. I noticed that if i send (writeAndFlush) tiny data on the channel isWritable continue to be true. But if I send about 839748 bytes on the channel the channel.isWritable remains false. I have taken heap dump in which i see that ChannelOutboundBuffer's buffer is unwritable. It has 12 bytes but zero buffer count; which is odd. 

<img width="1482" alt="screen shot 2016-10-17 at 7 32 29 pm" src="https://cloud.githubusercontent.com/assets/1609639/19460483/becdb46e-94a1-11e6-9509-2996c6893dc0.png">

<img width="1481" alt="screen shot 2016-10-17 at 7 31 55 pm" src="https://cloud.githubusercontent.com/assets/1609639/19460484/bed869cc-94a1-11e6-86b1-af6b46e6e05b.png">

I also noticed that io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged() now checks for invokeHandler() before firing the event. I wonder if that check is causing the problem. 

Could you please verify if this bug was recently added and possible fix for it?

**Comment by yogeshvedpathak** [comment]
update: I don't think that the issue is caused by change to io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged. That method was changed in 3.0.37 and the issue is only reproducible in 4.0.41. So it has to be something introduced in 4.0.41.

**Comment by yogeshvedpathak** [comment]
I figured out that the change d79e5c594d2e50168da6877e7d4de05733a404c7 broke the isWritable. It however affect if ChannelConfig.writeBufferLowWatermark is set to 0. If it is set to zero the condition newWriteBufferSize < channel.config().getWriteBufferLowWaterMark() in transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java does not match and channel never becomes writable again.


Proposed change to handle cases where low watermark is set to zero:

--- a/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java
+++ b/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java
@@ -194,7 +194,8 @@ public final class ChannelOutboundBuffer {
         }
 
         long newWriteBufferSize = TOTAL_PENDING_SIZE_UPDATER.addAndGet(this, -size);
-        if (notifyWritability && newWriteBufferSize < channel.config().getWriteBufferLowWaterMark()) {
+        if (notifyWritability && (newWriteBufferSize == 0
+           || newWriteBufferSize < channel.config().getWriteBufferLowWaterMark())) {
             setWritable(invokeLater);
         }
     }

**Comment by johnou** [comment]
@buchgr ping.

**Comment by johnou** [comment]
@yogeshvedpathak out of curiosity, why have you chosen a low watermark of 0?

**Comment by buchgr** [comment]
@johnou thanks for pointing me to this! :)

@yogeshvedpathak 

Thank you for digging in! Great job! Your are right, d79e5c5 changed the implementation to follow the API specification more closely [1]. The specification states that a channel is writable if the number of queued bytes `drops below the low write buffermark`. `0 < 0` is `false`, so the implementation seems correct.

However, I agree that this edge case is strange. Let me think about this for a bit.

**Comment by yogeshvedpathak** [comment]
no particular reason. I was just thinking that the actual values for low and high watermark do not matter but just the difference between them.

**Comment by buchgr** [comment]
@yogeshvedpathak so if you increase the low watermark the "bug" is gone, correct?

**Comment by almosthuman** [comment]
It seems as though if 0 isn't a valid low watermark, it shouldn't be permitted (the code checks that lowWaterMark is >=0, implying 0 is a valid value and it does make logical sense; in the case where you don't want the channel to be writable until the outbound buffer is empty, it would make sense to specify 0 rather than 1).  This seems like a case where perhaps it makes sense for the lower bound to be inclusive and the upper bound to be exclusive.

**Comment by yogeshvedpathak** [comment]
@buchgr yes setting low watermark to say, 1 resolves the issue.

**Comment by normanmaurer** [comment]
@buchgr ping

**Comment by buchgr** [comment]
> This seems like a case where perhaps it makes sense for the lower bound to be inclusive and the upper bound to be exclusive.

I agree, however that would in a sense be a breaking change :\.

As mentioned before, the current implementation seems to follow the specified interface. Is this still a problem @yogeshvedpathak ?

Maybe we should add an extra sentence to the documentation, explictly stating the implied meaning of `0`?

**Comment by normanmaurer** [comment]
@buchgr yep we can not break stuff... Please just add some more javadocs to make it clear.

**Comment by yogeshvedpathak** [comment]
After setting non-zero watermarks everything works fine for me. So a documentation change will suffice. Thanks.

**Comment by normanmaurer** [comment]
@buchgr ping :)

**Comment by normanmaurer** [comment]
@buchgr can you take care ?

**Comment by normanmaurer** [comment]
@buchgr ping... or should I take care ?

**Comment by buchgr** [comment]
@normanmaurer yes please take it over. I won't have time anytime soon. Sorry :\.
